ocrd_models/ocrd_models/ocrd_page_generateds.py

```diff
@@ -2850,6 +2850,43 @@ def buildChildren(self, child_, node, nodeName_, fromsubclass_=False, gds_collec
             obj_.original_tagname_ = 'CustomRegion'
     def __hash__(self):
         return hash(self.id)
+    def get_AllRegions(self, classes=None, order='document'):
+        """
+        Get all the *Region element or only those provided by ``classes``.
+        Returned in random order unless ``reading_order`` is set (NOT CURRENTLY IMPLEMENTED)
+        Arguments:
+            classes (list) Classes of regions that shall be returned, e.g. ['Text', 'Image']
+            order ("document"|"reading-order") Whether to return regions sorted by document order (default) or by reading order
+        """
+        if order not in ['document', 'reading-order']:
+            raise Exception("Argument 'order' must be either 'document' or 'reading-order', not '{}'".format(order))
+        if not classes:
+            classes = ['Advert', 'Chart', 'Chem', 'Custom', 'Graphic', 'Image', 'LineDrawing', 'Map', 'Maths', 'Music', 'Noise', 'Separator', 'Table', 'Text', 'Unknown']
+        def region_class(x):
+            return x.__class__.__name__.replace('RegionType', '')
+        ret = []
+        for region in classes:
+            ret += getattr(self, 'get_{}Region'.format(region))()
+        if order == 'reading-order':
+            reading_order = self.get_ReadingOrder().get_OrderedGroup() or reading_order.get_UnorderedGroup()
+            if reading_order:
+                def get_recursive_reading_order(rogroup):
+                    if isinstance(rogroup, (OrderedGroupType, OrderedGroupIndexedType)):
+                        elements = rogroup.get_AllIndexed()
+                    if isinstance(rogroup, (UnorderedGroupType, UnorderedGroupIndexedType)):
+                        elements = (rogroup.get_RegionRef() + rogroup.get_OrderedGroup() + rogroup.get_UnorderedGroup())
+                    regionrefs = list()
+                    for elem in elements:
+                        regionrefs.append(elem.get_regionRef())
+                        if not isinstance(elem, (RegionRefType, RegionRefIndexedType)):
+                            regionrefs.extend(get_recursive_reading_order(elem))
+                    return regionrefs
+                reading_order = get_recursive_reading_order(reading_order)
+            if reading_order:
+                id2region = dict([(region.id, region) for region in ret])
+                ret = [id2region[region_id] for region_id in reading_order if region_id in id2region]
```

Comment:
I see one further problem we did not think about before: What if not all regions are contained in `RegionOrder`? We could:
1. omit them from the result when running with `order='reading-order'`. (Current state of code), or
2. sort them at the beginning/end of the list, or
3. return them in an extra list (making the function two-valued), or
4. raise a (specialized) exception, returning both lists as members so this can be handled by the caller.

Reply:
@wrznr what would you say is most useful for writing clean and concise processors?

Reply:
I have extended the `order` parameter to allow both solution 1 (now: `reading-order-only`) and solution 2 (`reading-order`, with non-reading-order regions added to the end of the list).

Reply:
That's even better – thanks!